### PR TITLE
feat(router): Added public access to current instruction

### LINKED ISSUE #5335: Add possiblilty to ask router for the current route.
At the moment it is only possible to get the current `path` via the `Location.path()`method. When working with the "Routing DSL" it would be way more useful to check if a route is active via this DSL also. E.g. I want to do something like this

```
<li [class.active]="isActive('Dashboard')">
       <a [router-link]="['/Dashboard']" class="link">Dashboard</a>
</li>
```

But to be able to implement "isActive" I need the `name`of the current route. There are also cases where it would be useful to access the routeData of the current Route. At the moment it is only possible to access this information within the routed component, but not via the "Router" class

### PR DESCRIPTION
**Pull request by brandonroberts**
This provides public access to the current instruction and exposes the root router on each child router.

cc: @btford @petebacondarwin

### CODE REVIEW

**Comment by petebacondarwin** [comment]
This is a reasonable idea. What do you think @btford ?

**Comment by brandonroberts** [comment]
Finally got a green build on this one

**Comment by petebacondarwin** [comment]
I would like to think through the various use cases for accessing the complete instruction/URL/route information before agreeing to merge this...

**Comment by brandonroberts** [comment]
One use case is building a breadcrumb component. There isn't an easier way to get the instruction tree from a top down approach without accessing private properties. One way to get the full instruction is to subscribe to the router, get the url and resolve `router.recognize` promise to turn that url into the instruction.

**Comment by brandonroberts** [comment]
Another use case is reloading the current route programmatically. If you try to use `router.renavigate`, it doesn't work unless you used `router.navigateByUrl` previously. If that's the intended behavior, you could use `router.navigateByInstruction(router.currentInstruction)` to reload the current route.

**Comment by petebacondarwin** [comment]
I don't believe that we need this change for the use cases defined in #5335

But I accept @brandonroberts' first use case for building a breadcrumb tree. But I wonder if we need to make an API distinction between the "relative" current instruction, which is relative to the current router, and the "absolute" current instruction, which is relative to the root router?

In the second use case, I believe that the `renavigate()` method is supposed to be used when the route configuration may have changed and that the URL may no longer correspond to the current instruction.  So I don't think that you would use it for arbitrary reloading of routes.

I am not sure what the real world use case for reloading the current instruction is... Can you give one?

**Comment by brandonroberts** [comment]
@petebacondarwin I think a relative current instruction could be useful. My issue is that the router isn't going to force reload its parent routes and unless told not to, it will use reuse them. We already have the current instruction that are provided with the lifecycle hooks. Granted you can't use those for navigation and you can't use them to walk back up the route tree, but they are there.

One use case would be to force reload the entire ui for the current route after logging in through a modal. A user goes to a page that is publicly accessible but only shows limited information until the user logs in. The user triggers an action to login, a modal is displayed, they login and the entire route which is composed of parent/child routes is refreshed.

**Comment by petebacondarwin** [comment]
So what you are saying is that the resolution of a particular URL/route-instruction could be dependent upon information that is not contained in the URL/route-instruction such as whether the user is logged in.

**Comment by brandonroberts** [comment]
Yes

**Comment by petebacondarwin** [comment]
OK, so there are two goals:

1) support breadcrumb generation.
2) support forcing a reload of the current instruction.

The first is fixed by this PR - by accessing the top level instruction, which would allow the developer to traverse down the instructions - but I think we should call the method something slightly different to emphasize that it is getting the complete instruction from the root.  I was considering if we should expose the rootRouter directly on each router, then just expose the currentInstruction publicly. To get the full instruction we could then do `router.root.currentInstruction`. What do you think @brandonroberts ?

The second is not even fixable at the moment by this PR, since calling `navigateByInstruction` will just reuse the current components unless they explicitly prevent reuse by implementing `$canReuse` and so they may not automatically update their state... Again what do you think @brandonroberts ?

**Comment by brandonroberts** [comment]
With the first one, I'm fine with calling the method something different and exposing the current instruction. I think what you have proposed would be easier than traversing up the `router.parent` to get to the root. Would you consider making the RootRouter injectable like you did with ng1 component router? I don't know if that would be confusing to be able to inject the components router and the root router, but they serve different purposes.

Agreed on the second issue as well.

**Comment by brandonroberts** [comment]
@petebacondarwin I updated this PR to include the changes we discussed to access the root router and make the current instruction public.

**Comment by MonkeyDo** [comment]
Thanks @brandonroberts, that looks nice and easy.

**Comment by petebacondarwin** [comment]
I'll take a look later today!!

File: modules/angular2/src/router/router.ts [hunk 1]
```diff
@@ -501,6 +510,8 @@ class ChildRouter extends Router {
     // Delegate navigation to the root router
     return this.parent.navigateByInstruction(instruction, _skipLocationChange);
   }
+
+  get root() { return this.parent.root; }
```
Comment (petebacondarwin) [inline modules/angular2/src/router/router.ts]: I don't think this has to be generated here, right?
Can't we just have a straight attribute which is assigned in the constructor?

```
class RootRouter {
  constructor() {
    this.root = this;
  }
}
```

```
class ChildRouter {
  constructor(parent) {
    this.root = parent.root;
  }
}
```

**Comment by petebacondarwin** [comment]
One small note but I think the general idea is good.
I think we would need @btford or another Googler to chime in before we go ahead and modify the public API.

**Comment by brandonroberts** [comment]
@petebacondarwin good idea. I made the change you suggested.

File: modules/angular2/src/router/router.ts [hunk 2]
```diff
@@ -52,7 +55,6 @@ export class Router {
 
   constructor(public registry: RouteRegistry, public parent: Router, public hostComponent: any) {}
```
Comment (petebacondarwin) [inline modules/angular2/src/router/router.ts]: Sorry to be a pain but we can make providing the root even more explicit by including it as a constructor parameter here:

```ts
constructor(public registry: RouteRegistry, public parent: Router, public root: Router, public hostComponent: any) {}
```

Then we can ditch the declaration further up.

Comment (brandonroberts) [inline modules/angular2/src/router/router.ts]: No problem. I moved the root property to the end of the constructor and made it optional to satisfy TypeScript. The ROUTER_PROVIDERS provides the root router, and it won't have a root.

```
  constructor(public registry: RouteRegistry, public parent: Router, public hostComponent: any,
              public root?: Router) {}
```

Is that ok?

**Comment by petebacondarwin** [comment]
Great. Let's get @IgorMinar  or @mhevery  to review the API change.

**Comment by mhevery** [comment]
API change is fine.

**Comment by brandonroberts** [comment]
Cool. I got an error on the build but it wasn't caused by my changes.

**Comment by petebacondarwin** [comment]
It was a flake! I reran the build and it is green now.

**Comment by brandonroberts** [comment]
Thanks!

**Comment by vikerman** [comment]
TAP Green

**Comment by vikerman** [comment]
Manually merged to master.

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166459.
------------------------------------------------------------------------------------------------------------


[[8e95021111f09ee36e9c7e0a929b7d63e4c39ed6-673073415]]

**Comment by DrewynV** [comment]
__________________________________

Hi! We just wanted to let you know we receievd your email. Thank you for submitting your request. Our aim is to get back to you within 24 hours at the latest. If you have any additional information to add to your request, please reply to this email.


Also please visit our knowledgebase (http://www.support.magicdust.com.au) to browse for further info, we are adding to it all the time.

Or here is our industry report about the 11 Essentials for a Successful Website (http://www.magicdust.com.au/web-design/.

We look forward to helping you succeed online.

Thanks

The Team @ Magicdust Digital

http://support.magicdust.com.au/

Ph (02) 9979 7474
M-F | 9 am-5:00 pm AEST (AEDT in Summer) excluding Public Holidays

------------------------------------------------------------------------------------------------------------
This message was sent to  reply@reply.github.com in reference to Case #166508.
------------------------------------------------------------------------------------------------------------


[[39b5511997cd61e483d6249e93a6125b71dc43e0-673073915]]

**Pull request closed without merging**